File: android/src/main/java/com/cloudwebrtc/webrtc/FlutterDataPacketCryptor.java
package com.cloudwebrtc.webrtc;

import androidx.annotation.NonNull;

import com.cloudwebrtc.webrtc.utils.ConstraintsMap;

import org.webrtc.DataPacketCryptor;
import org.webrtc.DataPacketCryptorFactory;
import org.webrtc.FrameCryptor;
import org.webrtc.FrameCryptorAlgorithm;
import org.webrtc.FrameCryptorKeyProvider;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

class FlutterDataPacketCryptor {
    private static final String TAG = "FlutterDataPacketCryptor";
    private final Map<String, DataPacketCryptor> dataCryptos = new HashMap<>();

    private final FlutterRTCFrameCryptor frameCryptor;

    public FlutterDataPacketCryptor(FlutterRTCFrameCryptor frameCryptor) {
        this.frameCryptor = frameCryptor;
    }

    public boolean handleMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        String method_name = call.method;
        Map<String, Object> params = (Map<String, Object>) call.arguments;
        if (method_name.equals("createDataPacketCryptor")) {
            createDataPacketCryptor(params, result);
        } else if(method_name.equals("dataPacketCryptorEncrypt")) {
            dataPacketCryptorEncrypt(params, result);
        } else if(method_name.equals("dataPacketCryptorDecrypt")) {
            dataPacketCryptorDecrypt(params, result);
        } else if(method_name.equals("dataPacketCryptorDispose")) {
            dataPacketCryptorDispose(params, result);
        } else {
            return false;
        }
        return true;
    }

    private void createDataPacketCryptor(@NonNull Map<String, Object> params, @NonNull MethodChannel.Result result) {
        String keyProviderId = (String) params.get("keyProviderId");
        FrameCryptorKeyProvider keyProvider = frameCryptor.getKeyProvider(keyProviderId);
        if (keyProvider == null) {
            result.error("createDataPacketCryptorFailed", "keyProvider not found", null);
            return;
        }

        if(params.get("algorithm") == null) {
            result.error("createDataPacketCryptorFailed", "algorithm is null", null);
            return;
        }

        int algorithm = (int) params.get("algorithm");

        DataPacketCryptor dataPacketCryptor = DataPacketCryptorFactory.createDataPacketCryptor(
                frameCryptor.frameCryptorAlgorithmFromInt(algorithm),
                keyProvider);
        if(dataPacketCryptor == null) {
            result.error("createDataPacketCryptorFailed", "createDataPacketCryptor failed", null);
            return;
        }

        String dataCryptorId = UUID.randomUUID().toString();
        dataCryptos.put(dataCryptorId, dataPacketCryptor);

        ConstraintsMap paramsResult = new ConstraintsMap();
        paramsResult.putString("dataCryptorId", dataCryptorId);
        result.success(paramsResult.toMap());
    }

    private void dataPacketCryptorEncrypt(@NonNull Map<String, Object> params, @NonNull MethodChannel.Result result) {
        String dataCryptorId = (String) params.get("dataCryptorId");
        if (dataCryptorId == null) {
            result.error("dataPacketCryptorEncryptFailed", "dataCryptorId is null", null);
            return;
        }

        DataPacketCryptor dataPacketCryptor = dataCryptos.get(dataCryptorId);
        if(dataPacketCryptor == null) {
            result.error("dataPacketCryptorEncryptFailed", "dataPacketCryptor not found", null);
            return;
        }

        String participantId = (String) params.get("participantId");
        if (participantId == null) {
            result.error("dataPacketCryptorEncryptFailed", "participantId is null", null);
            return;
        }

        byte[] data = (byte[]) params.get("data");
        if (data == null) {
            result.error("dataPacketCryptorEncryptFailed", "data is null", null);
            return;
        }

        int keyIndex = (int) params.get("keyIndex");
        if( keyIndex < 0 ) {
            result.error("dataPacketCryptorEncryptFailed", "keyIndex is invalid", null);
            return;
        }

        DataPacketCryptor.EncryptedPacket packet = dataPacketCryptor.encrypt(participantId, keyIndex, data);
        ConstraintsMap paramsResult = new ConstraintsMap();
        paramsResult.putInt("keyIndex", packet.keyIndex);
        paramsResult.putByte("data", packet.payload);
        paramsResult.putByte("iv", packet.iv);
        result.success(paramsResult.toMap());
    }

    private void dataPacketCryptorDecrypt(@NonNull Map<String, Object> params, @NonNull MethodChannel.Result result) {
        String dataCryptorId = (String) params.get("dataCryptorId");
        if (dataCryptorId == null) {
            result.error("dataPacketCryptorEncryptFailed", "dataCryptorId is null", null);
            return;
        }

        DataPacketCryptor dataPacketCryptor = dataCryptos.get(dataCryptorId);
        if(dataPacketCryptor == null) {
            result.error("dataPacketCryptorEncryptFailed", "dataPacketCryptor not found", null);
            return;
        }

        String participantId = (String) params.get("participantId");
        if (participantId == null) {
            result.error("dataPacketCryptorEncryptFailed", "participantId is null", null);
            return;
        }

        byte[] data = (byte[]) params.get("data");
        if (data == null) {
            result.error("dataPacketCryptorEncryptFailed", "data is null", null);
            return;
        }

        byte[] iv = (byte[]) params.get("iv");
        if (iv == null) {
            result.error("dataPacketCryptorEncryptFailed", "iv is null", null);
            return;
        }

        int keyIndex = (int) params.get("keyIndex");
        if( keyIndex < 0 ) {
            result.error("dataPacketCryptorEncryptFailed", "keyIndex is invalid", null);
            return;
        }
        DataPacketCryptor.EncryptedPacket encryptedPacket = new DataPacketCryptor.EncryptedPacket(data, iv, keyIndex);
        byte[] decrypted = dataPacketCryptor.decrypt(participantId, encryptedPacket);
        if(decrypted == null) {
            result.error("dataPacketCryptorDecryptFailed", "decrypt failed", null);
            return;
        }

        ConstraintsMap paramsResult = new ConstraintsMap();
        paramsResult.putByte("data", decrypted);
        result.success(paramsResult.toMap());
    }

    private void dataPacketCryptorDispose(@NonNull Map<String, Object> params, @NonNull MethodChannel.Result result) {
        String dataCryptorId = (String) params.get("dataCryptorId");
        if (dataCryptorId == null) {
            result.error("dataPacketCryptorDisposeFailed", "dataCryptorId is null", null);
            return;
        }

        DataPacketCryptor dataPacketCryptor = dataCryptos.remove(dataCryptorId);
        if(dataPacketCryptor == null) {
            result.error("dataPacketCryptorDisposeFailed", "dataPacketCryptor not found", null);
            return;
        }

        if (dataPacketCryptor != null) {
            dataPacketCryptor.dispose();
        }
        result.success(null);
    }
}

File: common/darwin/Classes/FlutterDataPacketCryptor.m
#import "FlutterDataPacketCryptor.h"
#import "FlutterRTCFrameCryptor.h"

#include <WebRTC/WebRTC.h>
#import <objc/runtime.h>

@implementation FlutterWebRTCPlugin (DataPacketCryptor)

- (void)handleDataPacketCryptorMethodCall:(nonnull FlutterMethodCall*)call
                                   result:(nonnull FlutterResult)result {
  NSDictionary* constraints = call.arguments;
  NSString* method = call.method;
  if ([method isEqualToString:@"createDataPacketCryptor"]) {
    [self createDataPacketCryptor:constraints result:result];
  } else if ([method isEqualToString:@"dataPacketCryptorEncrypt"]) {
    [self dataPacketCryptorEncrypt:constraints result:result];
  } else if ([method isEqualToString:@"dataPacketCryptorDecrypt"]) {
    [self dataPacketCryptorDecrypt:constraints result:result];
  } else if ([method isEqualToString:@"dataPacketCryptorDispose"]) {
    [self dataPacketCryptorDispose:constraints result:result];
  } else {
    result(FlutterMethodNotImplemented);
  }
}

- (void)createDataPacketCryptor:(nonnull NSDictionary*)constraints
                         result:(nonnull FlutterResult)result {
  NSNumber* algorithm = constraints[@"algorithm"];
  if (algorithm == nil) {
    result([FlutterError errorWithCode:@"createDataPacketCryptorFailed"
                               message:@"Invalid algorithm"
                               details:nil]);
    return;
  }

  NSString* keyProviderId = constraints[@"keyProviderId"];
  if (keyProviderId == nil) {
    result([FlutterError errorWithCode:@"createDataPacketCryptorFailed"
                               message:@"Invalid keyProviderId"
                               details:nil]);
    return;
  }

  RTCFrameCryptorKeyProvider* keyProvider = self.keyProviders[keyProviderId];
  if (keyProvider == nil) {
    result([FlutterError errorWithCode:@"createDataPacketCryptorFailed"
                               message:@"Invalid keyProvider"
                               details:nil]);
    return;
  }

  RTCDataPacketCryptor* dataPacketCryptor =
      [[RTCDataPacketCryptor alloc] initWithAlgorithm:[self getAlgorithm:algorithm]
                                          keyProvider:keyProvider];
  NSString* dataCryptorId = [[NSUUID UUID] UUIDString];
  self.dataCryptors[dataCryptorId] = dataPacketCryptor;

  result(@{@"dataCryptorId" : dataCryptorId});
}

- (void)dataPacketCryptorDispose:(nonnull NSDictionary*)constraints
                          result:(nonnull FlutterResult)result {
  NSString* dataCryptorId = constraints[@"dataCryptorId"];
  if (dataCryptorId == nil) {
    result([FlutterError errorWithCode:@"dataPacketCryptorDisposeFailed"
                               message:@"Invalid dataCryptorId"
                               details:nil]);
    return;
  }

  RTCDataPacketCryptor* dataPacketCryptor = self.dataCryptors[dataCryptorId];
  if (dataPacketCryptor == nil) {
    result([FlutterError errorWithCode:@"dataPacketCryptorDisposeFailed"
                               message:@"Invalid dataCryptor"
                               details:nil]);
    return;
  }

  [self.dataCryptors removeObjectForKey:dataCryptorId];
  result(nil);
}

- (void)dataPacketCryptorEncrypt:(nonnull NSDictionary*)constraints
                          result:(nonnull FlutterResult)result {
  NSString* dataCryptorId = constraints[@"dataCryptorId"];
  if (dataCryptorId == nil) {
    result([FlutterError errorWithCode:@"dataPacketCryptorEncryptFailed"
                               message:@"Invalid dataCryptorId"
                               details:nil]);
    return;
  }

  FlutterStandardTypedData* data = constraints[@"data"];
  if (data == nil) {
    result([FlutterError errorWithCode:@"dataPacketCryptorEncryptFailed"
                               message:@"Invalid data"
                               details:nil]);
    return;
  }

  NSString* participantId = constraints[@"participantId"];
  if (participantId == nil) {
    result([FlutterError errorWithCode:@"dataPacketCryptorEncryptFailed"
                               message:@"Invalid iv"
                               details:nil]);
    return;
  }

  NSNumber* keyIndex = constraints[@"keyIndex"];
  if (keyIndex == nil) {
    result([FlutterError errorWithCode:@"dataPacketCryptorEncryptFailed"
                               message:@"Invalid keyIndex"
                               details:nil]);
    return;
  }

  RTCDataPacketCryptor* dataPacketCryptor = self.dataCryptors[dataCryptorId];
  if (dataPacketCryptor == nil) {
    result([FlutterError errorWithCode:@"dataPacketCryptorEncryptFailed"
                               message:@"Invalid dataCryptor"
                               details:nil]);
    return;
  }

  RTC_OBJC_TYPE(RTCEncryptedPacket)* encryptedData =
      [dataPacketCryptor encrypt:participantId keyIndex:keyIndex.unsignedIntValue data:data.data];
  if (encryptedData == nil) {
    result([FlutterError errorWithCode:@"dataPacketCryptorEncryptFailed"
                               message:@"Encrypt failed"
                               details:nil]);
    return;
  }

  result(@{
    @"data" : encryptedData.data,
    @"iv" : encryptedData.iv,
    @"keyIndex" : @(encryptedData.keyIndex)
  });
}

- (void)dataPacketCryptorDecrypt:(nonnull NSDictionary*)constraints
                          result:(nonnull FlutterResult)result {
  NSString* dataCryptorId = constraints[@"dataCryptorId"];
  if (dataCryptorId == nil) {
    result([FlutterError errorWithCode:@"dataPacketCryptorDecryptFailed"
                               message:@"Invalid dataCryptorId"
                               details:nil]);
    return;
  }

  RTCDataPacketCryptor* dataPacketCryptor = self.dataCryptors[dataCryptorId];
  if (dataPacketCryptor == nil) {
    result([FlutterError errorWithCode:@"dataPacketCryptorDecryptFailed"
                               message:@"Invalid dataCryptor"
                               details:nil]);
    return;
  }

  FlutterStandardTypedData* data = constraints[@"data"];
  if (data == nil) {
    result([FlutterError errorWithCode:@"dataPacketCryptorDecryptFailed"
                               message:@"Invalid data"
                               details:nil]);
    return;
  }

  FlutterStandardTypedData* iv = constraints[@"iv"];
  if (iv == nil) {
    result([FlutterError errorWithCode:@"dataPacketCryptorDecryptFailed"
                               message:@"Invalid iv"
                               details:nil]);
    return;
  }

  NSNumber* keyIndex = constraints[@"keyIndex"];
  if (keyIndex == nil) {
    result([FlutterError errorWithCode:@"dataPacketCryptorDecryptFailed"
                               message:@"Invalid keyIndex"
                               details:nil]);
    return;
  }

  NSString* participantId = constraints[@"participantId"];
  if (participantId == nil) {
    result([FlutterError errorWithCode:@"dataPacketCryptorDecryptFailed"
                               message:@"Invalid participantId"
                               details:nil]);
    return;
  }

  RTCEncryptedPacket* encryptedPacket =
      [[RTCEncryptedPacket alloc] initWithData:data.data
                                            iv:iv.data
                                      keyIndex:keyIndex.unsignedIntValue];
  NSData* decryptedData = [dataPacketCryptor decrypt:participantId encryptedPacket:encryptedPacket];
  if (decryptedData == nil) {
    result([FlutterError errorWithCode:@"dataPacketCryptorDecryptFailed"
                               message:@"Decrypt failed"
                               details:nil]);
    return;
  }

  result(@{@"data" : decryptedData});
}

@end
